File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/serializeAndDeserialize.ts
import {arrayToBTLevelOrder, BTNode, levelOrder} from "./BinaryTreeOperations";

function serialize(root: BTNode) {
	if (!root) return "";

	let s = "";
	let q = [];
	q.push(root);
	while (q.length > 0) {
		let curNode = q.shift();
		if (curNode === null) {
			s += "#,";
		} else {
			s += curNode.data + ",";
			q.push(curNode.left);
			q.push(curNode.right);
		}
	}
	return s;
}

function deserialize(data: string) {
	if (data.length === 0) return null;
	let dataArray = data.split(",");
	let root = new BTNode(parseInt(dataArray[0]));
	let q = [];
	q.push(root);
	let i = 1;
	while (i < dataArray.length) {
		let node = q.shift();
		let leftVal = dataArray[i++];
		if (leftVal !== "#" && leftVal) {
			let leftNode = new BTNode(parseInt(leftVal));
			node.left = leftNode;
			q.push(leftNode);
		}
		let rightVal = dataArray[i++];
		if (rightVal !== "#" && rightVal) {
			let rightNode = new BTNode(parseInt(rightVal));
			node.right = rightNode;
			q.push(rightNode);
		}
	}
	return root;
}

let root = arrayToBTLevelOrder([1,2,3,4,5,6,7,8])
let s = serialize(root);
let root1 = deserialize(s);
levelOrder(root1);

File: StriverA2Z/Java/BinaryTrees/SerializeAndDeserialize.java
import java.util.*;

public class SerializeAndDeserialize {
    public static void main(String[] args) {
        Node root = BinaryTreeOperations.arrayToBTLevelOrder(new Integer[] {1,2,3,4,5,6,7,8});
        String s = serialize(root);
        Node root1 = deserialize(s);
        BinaryTreeOperations.levelOrder(root1);
    }
    public static String serialize(Node root) {
        if (root == null) return "";
        Queue<Node> q = new LinkedList<>();
        StringBuilder res = new StringBuilder();
        q.add(root);
        while (!q.isEmpty()) {
            Node node = q.poll();
            if (node == null) {
                res.append("n ");
                continue;
            }
            res.append(node.data).append(" ");
            q.add(node.left);
            q.add(node.right);
        }
        return res.toString();
    }

    public static Node deserialize(String data) {
        if (data == "") return null;
        Queue<Node> q = new LinkedList<>();
        String[] values = data.split(" ");
        Node root = new Node(Integer.parseInt(values[0]));
        q.add(root);
        for (int i = 1; i < values.length; i++) {
            Node parent = q.poll();
            if (!values[i].equals("n")) {
                Node left = new Node(Integer.parseInt(values[i]));
                parent.left = left;
                q.add(left);
            }
            if (!values[++i].equals("n")) {
                Node right = new Node(Integer.parseInt(values[i]));
                parent.right = right;
                q.add(right);
            }
        }
        return root;
    }
}
